File: spanner/src/main/java/papyrus/spanner/Spanner.java
package papyrus.spanner;

import android.text.Spannable;
import android.text.SpannableString;

import java.util.ArrayList;
import java.util.List;

public class Spanner {

    public static class Builder {
        StringBuilder stringBuilder = new StringBuilder();
        List<SpanDeclaration> spanRegistry = new ArrayList<>();

        public Builder appendSpan(String text, Object... spans) {
            for (Object obj : spans) {
                spanRegistry.add(new SpanDeclaration(obj, stringBuilder.length(), stringBuilder.length() + text.length()));
            }
            stringBuilder.append(text);
            return this;
        }

        public Spannable build() {
            SpannableString str = new SpannableString(stringBuilder.toString());
            for (SpanDeclaration span : spanRegistry) {
                str.setSpan(span.what, span.start, span.end, SpannableString.SPAN_INCLUSIVE_EXCLUSIVE);
            }

            return str;
        }

    }

    private static class SpanDeclaration {
        Object what;
        int start;
        int end;

        public SpanDeclaration(Object what, int start, int end) {
            this.what = what;
            this.start = start;
            this.end = end;
        }
    }
}

File: spanner/src/main/java/papyrus/spanner/Spanner.kt
package papyrus.spanner

import android.os.Parcel
import android.os.Parcelable
import android.text.Spannable
import android.text.SpannableString
import papyrus.core.navigation.readArbitrary
import papyrus.core.navigation.writeArbitrary
import java.util.*

@Suppress("UNCHECKED_CAST")
class Spanner() : Parcelable {
    private var stringBuilder = StringBuilder()
    private var spanRegistry = ArrayList<SpanDeclaration>()
    private var spanStack = Stack<SpanDeclaration>()

    constructor(parcel: Parcel) : this() {
        stringBuilder = parcel.readSerializable() as StringBuilder
        spanRegistry = parcel.readArbitrary()
    }

    fun beginSpan(obj: Any): Spanner {
        spanStack.push(SpanDeclaration(obj, stringBuilder.length, stringBuilder.length))
        return this
    }

    fun appendSpan(text: String, vararg spans: Any): Spanner {
        spans.forEach { obj ->
            spanRegistry.add(SpanDeclaration(obj, stringBuilder.length, stringBuilder.length + text.length))
        }
        stringBuilder.append(text)
        return this
    }

    fun endSpan(): Spanner {
        if (spanStack.isNotEmpty()) {
            spanStack.pop()?.apply {
                end = stringBuilder.length
            }?.let(spanRegistry::add)
        }
        return this
    }

    fun build(): Spannable = SpannableString(stringBuilder.toString()).apply {
        spanRegistry.forEach { span ->
            span.what?.let { obj ->
                setSpan(obj, span.start, span.end, SpannableString.SPAN_INCLUSIVE_EXCLUSIVE)
            }
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeSerializable(stringBuilder)
        parcel.writeArbitrary(spanRegistry)
    }

    override fun describeContents(): Int = 0

    companion object CREATOR : Parcelable.Creator<Spanner> {
        override fun createFromParcel(parcel: Parcel): Spanner {
            return Spanner(parcel)
        }

        override fun newArray(size: Int): Array<Spanner?> {
            return arrayOfNulls(size)
        }
    }
}

data class SpanDeclaration(var what: Any?, var start: Int, var end: Int)
